From 2489f91d395740b42f21322744ab64427e94a683 Mon Sep 17 00:00:00 2001
From: Jiffin Tony Thottan <jthottan@redhat.com>
Date: Tue, 31 Jul 2018 14:52:51 +0530
Subject: [PATCH 346/351] gfapi : Handle the path == "" glfs_resolve_at

Currently there is no check for path = "" in glfs_resolve_at.
So if application sends an empty path, then the function resolves
into the parent inode which is incorrect. Plus modified possible
of "path" with "origpath" in the same function.

Upstream reference
>Change-Id: Ie5ff9ce4b771607b7dbb3fe00704fe670421792a
>fixes: bz#1610236
>Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>
>link : https://review.gluster.org/#/c/glusterfs/+/20594/

Change-Id: Ie5ff9ce4b771607b7dbb3fe00704fe670421792a
Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/146938
Reviewed-by: Sunil Kumar Heggodu Gopala Acharya <sheggodu@redhat.com>
---
 api/src/glfs-resolve.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/api/src/glfs-resolve.c b/api/src/glfs-resolve.c
index 76835cb..95fe67a 100644
--- a/api/src/glfs-resolve.c
+++ b/api/src/glfs-resolve.c
@@ -472,14 +472,13 @@ priv_glfs_resolve_at (struct glfs *fs, xlator_t *subvol, inode_t *at,
 	DECLARE_OLD_THIS;
 	__GLFS_ENTRY_VALIDATE_FS(fs, invalid_fs);
 
-	path = gf_strdup (origpath);
-	if (!path) {
-		errno = ENOMEM;
-		return -1;
-	}
+        if (origpath[0] == '\0') {
+                errno = EINVAL;
+                goto invalid_fs;
+        }
 
 	parent = NULL;
-	if (at && path[0] != '/') {
+	if (at && origpath[0] != '/') {
 		/* A relative resolution of a path which starts with '/'
 		   is equal to an absolute path resolution.
 		*/
@@ -487,10 +486,14 @@ priv_glfs_resolve_at (struct glfs *fs, xlator_t *subvol, inode_t *at,
 	} else {
 		inode = inode_ref (subvol->itable->root);
 
-		if (strcmp (path, "/") == 0)
+		if (strcmp (origpath, "/") == 0)
                         glfs_resolve_root (fs, subvol, inode, &ciatt);
 	}
 
+        path = gf_strdup (origpath);
+	if (!path)
+		goto invalid_fs;
+
 	for (component = strtok_r (path, "/", &saveptr);
 	     component; component = next_component) {
 
-- 
1.8.3.1

